binding.cc

```diff
@@ -249,6 +251,37 @@ static size_t StringOrBufferLength (napi_env env, napi_value value) {
   return size;
 }
 
+struct NapiSlice {
+  NapiSlice(napi_env env, napi_value from) {
+    if (IsString(env, from)) {
+      NAPI_STATUS_THROWS_VOID(napi_get_value_string_utf8(env, from, nullptr, 0, &size_));
+      char* data;
+      if (size_ + 1 < stack_.size()) {
+        data = stack_.data();
+      } else {
+        heap_.reset(new char[size_ + 1]);
+        data = heap_.get();
+      }
+      data[size_] = 0;
+      NAPI_STATUS_THROWS_VOID(napi_get_value_string_utf8(env, from, data, size_ + 1, &size_));
+      data_ = data;
+    } else if (IsBuffer(env, from)) {
+      void* data;
+      NAPI_STATUS_THROWS_VOID(napi_get_buffer_info(env, from, &data, &size_));
+      data_ = static_cast<char*>(data);
+    }
+  }
+
+  operator leveldb::Slice() const {
+    return leveldb::Slice(data_, size_);
+  }
+
+  std::unique_ptr<char[]> heap_;
+  std::array<char, 1024> stack_;
```

Comment:
This looks like an optimization for an overly-specific use case.

Reply:
Most keys are < 1024 in length...

Reply:
That's probably a fair assumption, but doesn't it then over-allocate memory?

Reply:
it's stack allocated?

Reply:
Maybe 1024 is a bit high considering cache locality...